src/compaction_filter.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+#include "db_impl.h"
+#include "rocksdb/compaction_filter.h"
+
+#include <utility>
+
+namespace rocksdb {
+namespace titandb {
+
+class TitanCompactionFilter final : public CompactionFilter {
+public:
+  explicit TitanCompactionFilter(TitanDBImpl *db,
+                                 const CompactionFilter *original,
+                                 std::shared_ptr<BlobStorage> blob_storage,
+                                 bool manual_release)
+      : db_(db), blob_storage_(std::move(blob_storage)),
+        original_filter_(original), manual_release_(manual_release) {}
+
+  ~TitanCompactionFilter() override {
+    if (manual_release_) {
+      delete original_filter_;
+    }
+  }
+
+  const char *Name() const override {
+    return std::string("TitanCompactionFilter.")
+        .append(original_filter_->Name())
+        .c_str();
+  }
+
+  Decision FilterV2(int level, const Slice &key, ValueType value_type,
+                    const Slice &value, std::string *new_value,
+                    std::string *skip_until) const override {
+    if (value_type != kBlobIndex) {
+      return original_filter_->FilterV2(level, key, value_type, value,
+                                        new_value, skip_until);
+    }
+
+    BlobIndex blob_index;
+    Slice original_value(value.data());
+    Status s = blob_index.DecodeFrom(&original_value);
+    if (!s.ok()) {
+      db_->SetBGError(s);
```

Comment:
print an info log using `ROCKS_LOG_ERROR` macro? you will need to pass `db_->db_options_->info_log` into `TitanCompactionFilter`.